kube/client.go

```diff
@@ -148,41 +160,183 @@ func isCompatible(clientMajor, clientMinor, serverMajor, serverMinor string) err
+	if c.Server != "" {
+		args = append(args, fmt.Sprintf("--kubeconfig=%s", kubeconfigFilePath))
+	}
+
+	return executeApply(args)
+}
+
+// StrictApply will attempt to "kubectl apply" the file located at path using a `kube-applier` service account under the given namespace.
+// `kube-applier` service account must exist for the given namespace and must contain a secret that include token and ca.cert.
+// It returns the full apply command and its output.
+func (c *Client) StrictApply(path, namespace string, dryRun bool) (string, string, error) {
+	args := prepareApplyArgs(path, namespace, c.Label, dryRun)
+
+	tempKubeConfigFilepath, tempCertFilepath, err := c.CreateTempConfig(namespace, "kube-applier")
+	if err != nil {
+		return "", "", fmt.Errorf("error creating temp config: %v", err)
+	}
+	defer func() { os.Remove(tempKubeConfigFilepath); os.Remove(tempCertFilepath) }()
+	args = append(args, fmt.Sprintf("--kubeconfig=%s", tempKubeConfigFilepath))
+
+	return executeApply(args)
+}
+
+func decodeToJson(i interface{}, raw_message []byte) error {
+	dec := json.NewDecoder(bytes.NewReader(raw_message))
+
+	if err := dec.Decode(i); err != nil {
+		return fmt.Errorf("Response is not json format: %v error=(%v)", string(raw_message), err)
+	}
+
+	return nil
+}
+
 func (c *Client) GetNamespaceStatus(namespace string) (AutomaticDeploymentOption, error) {
 	args := []string{"kubectl", "get", "namespace", namespace, "-o", "json", "-n", namespace}
 	if c.Server != "" {
 		args = append(args, fmt.Sprintf("--kubeconfig=%s", kubeconfigFilePath))
 	}
-	stdout, err := exec.Command(args[0], args[1:]...).CombinedOutput()
+	stdout, err := execCommand(args[0], args[1:]...).CombinedOutput()
 	if err != nil {
 		return Off, err
 	}
-	dec := json.NewDecoder(bytes.NewReader(stdout))
+
 	var nr struct {
 		Metadata struct {
 			Labels map[string]string
 		}
 	}
-	if err := dec.Decode(&nr); err != nil {
-		return Off, fmt.Errorf("Get namespace response is not json format: %v error=(%v)", string(stdout), err)
+	if err := decodeToJson(&nr, stdout); err != nil {
+		return Off, err
 	}
+
 	return AutomaticDeploymentOption(nr.Metadata.Labels[c.Label]), nil
 }
+
+func (c *Client) GetNamespaceUserSecretName(namespace, username string) (string, error) {
+	args := []string{"kubectl", "get", "serviceaccount", username, "-o", "json", "-n", namespace}
+	if c.Server != "" {
+		args = append(args, fmt.Sprintf("--kubeconfig=%s", kubeconfigFilePath))
+	}
+	stdout, err := execCommand(args[0], args[1:]...).CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+
+	type secret struct {
+		Name string
+	}
+	var nr struct {
+		Secrets []secret
+	}
+	if err := decodeToJson(&nr, stdout); err != nil {
+		return "", err
+	}
+
+	if len(nr.Secrets) > 1 {
+		log.Printf("Found many secrets for kube-applier on %s using the first on the list", namespace)
+	}
+	if len(nr.Secrets) == 0 {
+		return "", fmt.Errorf("No secrets found for %s user", username)
```

Comment:
Errors strings should not be capitalized
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Also, I think there is a linter for that, example config you could use: https://github.com/utilitywarehouse/go-circle-build/blob/master/Makefile#L24-L51